crates/telio-crypto/src/lib.rs

```diff
@@ -26,23 +26,60 @@ pub mod encryption;
 use std::{cmp::Ordering, convert::TryInto, fmt};
 
 use rand::prelude::*;
+use serde::{Deserialize, Serialize};
 use serde_with::{DeserializeFromStr, SerializeDisplay};
+use telio_utils::Hidden;
 
 /// Secret, Public and Wireguard Preshared key size in bytes
 pub const KEY_SIZE: usize = 32;
 
 /// Secret key type
 #[derive(
-    Default, PartialOrd, Ord, PartialEq, Eq, Hash, Copy, Clone, DeserializeFromStr, SerializeDisplay,
+    Default, Debug, PartialOrd, Ord, PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize,
 )]
-pub struct SecretKey([u8; KEY_SIZE]);
+pub struct SecretKey(Hidden<[u8; KEY_SIZE]>);
 
 /// Public key type
 #[derive(
     Default, PartialOrd, Ord, PartialEq, Eq, Hash, Copy, Clone, DeserializeFromStr, SerializeDisplay,
 )]
 pub struct PublicKey(pub [u8; KEY_SIZE]);
 
+impl fmt::Display for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut buf = [0u8; 44];
+        base64::encode_config_slice(self.0, base64::STANDARD, &mut buf);
+        match std::str::from_utf8(&buf) {
+            Ok(buf) => f.write_str(buf),
+            Err(_) => Err(fmt::Error),
+        }
+    }
+}
+
+impl fmt::Debug for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let buf = base64::encode(self.0);
+        f.write_str(&format!(
+            "\"{:.*}...{}\"",
+            4,
+            &buf,
+            &buf.get((buf.len()) - 4..).ok_or(fmt::Error)?
+        ))
+    }
+}
```

Comment:
I'm curious why `Display` prints the full key and `Debug` prints the reduced version rather than the other way around. Maybe I'm missing something smart?

Reply:
I don't know why was this chosen this way. What I did here, is move the code template which was in the `gen_common` and apply it only to the `PublicKey`.